src/main/java/io/burt/athena/polling/FixedDelayPollingStrategy.java

```diff
@@ -11,24 +13,34 @@
 public class FixedDelayPollingStrategy implements PollingStrategy {
     private final Duration delay;
     private final Sleeper sleeper;
+    private Clock clock;
 
     FixedDelayPollingStrategy(Duration delay) {
-        this(delay, duration -> TimeUnit.MILLISECONDS.sleep(duration.toMillis()));
+        this(delay, duration -> TimeUnit.MILLISECONDS.sleep(duration.toMillis()), Clock.systemDefaultZone());
     }
 
-    FixedDelayPollingStrategy(Duration delay, Sleeper sleeper) {
+    FixedDelayPollingStrategy(Duration delay, Sleeper sleeper, Clock clock) {
         this.delay = delay;
         this.sleeper = sleeper;
+        this.clock = clock;
     }
 
     @Override
-    public ResultSet pollUntilCompleted(PollingCallback callback) throws SQLException, TimeoutException, ExecutionException, InterruptedException {
+    public ResultSet pollUntilCompleted(PollingCallback callback, Instant deadline) throws SQLException, TimeoutException, ExecutionException, InterruptedException {
         while (true) {
-            Optional<ResultSet> resultSet = callback.poll();
+            Optional<ResultSet> resultSet = callback.poll(deadline);
             if (resultSet.isPresent()) {
                 return resultSet.get();
             } else {
-                sleeper.sleep(delay);
+                Duration beforeDeadline = Duration.between(clock.instant(), deadline);
+                if (beforeDeadline.compareTo(delay) < 0) {
+                    if (beforeDeadline.isNegative()) {
+                        throw new TimeoutException();
+                    }
+                    sleeper.sleep(beforeDeadline);
+                } else {
+                    sleeper.sleep(delay);
+                }
```

Comment:
This is the same code as in `BackoffPollingStrategy`, which is fine, but perhaps it could be either moved to a default method in the `PollingStrategy` interface, or to a static helper method somewhere? I suspect the next implementation would want to do more or less the same thing.

Something like `void pollSleep(Instant now, Instant deadline, Duration maxDelay) throws TimeoutException`.

Reply:
The sleeper isn't accessible in the base class, but extracted a `sleepDuration` method that I think matches your intent.

When I read this comment I also realized that I forgot to mention something about the polling. The way it works now, you get a final attempt from the polling strategies when there is no time remaining, but with how poll is actually implemented now, that won't be particularly helpful, as poll will always take some time, and a timeout would then typically happen when we wait zero seconds for an API operation to complete.

The alternatives I see would be to effectively shrink the deadline in `AthenaStatement` by a network timeout and then perform operations in `AthenaStatement` with the deadline plus the network timeout, but that feels a bit dirty. The other option would be for the polling strategies to know about the network timeout, but I'm not sure that makes sense.

Another option would obviously be to just timeout if we were to sleep for "too long", but that effectively means that timeouts are going to be a lot shorter than configured.